File: cpp/src/fedpro/utility/StandAloneMovingStats.cpp
/***********************************************************************
  Copyright (C) 2023 Pitch Technologies AB

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 **********************************************************************/

#include "StandAloneMovingStats.h"

#include <algorithm>
#include <chrono>
#include <limits>
#include <stdexcept>

namespace FedPro
{
   constexpr const FedProDuration StandAloneMovingStats::DEFAULT_MILLIS_PER_BUCKET;

   StandAloneMovingStats::StandAloneMovingStats(
         FedProDuration maxTimeWindowMillis)
         : StandAloneMovingStats(maxTimeWindowMillis, nowMillis())
   {
   }

   StandAloneMovingStats::StandAloneMovingStats(
         FedProDuration maxTimeWindowMillis,
         SteadyTimePoint startTime)
         : StandAloneMovingStats(maxTimeWindowMillis, DEFAULT_MILLIS_PER_BUCKET, startTime)
   {
   }

   StandAloneMovingStats::StandAloneMovingStats(
         FedProDuration maxTimeWindowMillis,
         FedProDuration millisPerBucket,
         SteadyTimePoint startTime)
         : _bucketCount((maxTimeWindowMillis.count() / millisPerBucket.count()) * 2),
           _millisPerBucket{millisPerBucket},
           _maxPerBucket(_bucketCount, std::numeric_limits<Sample>::min()),
           _minPerBucket(_bucketCount, std::numeric_limits<Sample>::max()),
           _sumPerBucket(_bucketCount, 0),
           _sampleCountPerBucket(_bucketCount, 0),
           _startTime{startTime}
   {
      if (maxTimeWindowMillis.count() / millisPerBucket.count() > std::numeric_limits<size_t>::max() / 2) {
         throw std::invalid_argument{
               "maxTimeWindowMillis/millisPerBucket is too large, would overflow the number of buckets. "};
      }
   }

   void StandAloneMovingStats::sample(int value)
   {
      sample(value, nowMillis());
   }

   void StandAloneMovingStats::sample(
         int value,
         SteadyTimePoint time)
   {
      size_t currentBucket{getBucket(time)};
      {
         std::lock_guard<std::mutex> lock(_guard);
         resetBucketsBeforeUsing(currentBucket);
         size_t currentBucketIndex = currentBucket % _bucketCount;
         _maxPerBucket[currentBucketIndex] = std::max(_maxPerBucket[currentBucketIndex], value);
         _minPerBucket[currentBucketIndex] = std::min(_minPerBucket[currentBucketIndex], value);
         _sumPerBucket[currentBucketIndex] += value;
         _sampleCountPerBucket[currentBucketIndex]++;
         _historicTotal += value;
         _historicCount++;
      }
   }

   StandAloneMovingStats::Stats StandAloneMovingStats::getStats()
   {
      return getStatsForTime(nowMillis());
   }

   StandAloneMovingStats::Stats StandAloneMovingStats::getStats(FedProDuration timeWindowMillis)
   {
      return getStatsForTime(timeWindowMillis, nowMillis());
   }

   StandAloneMovingStats::Stats StandAloneMovingStats::getStatsForTime(SteadyTimePoint time)
   {
      return getStatsForTime((_bucketCount / 2) * _millisPerBucket, time);
   }

   StandAloneMovingStats::Stats StandAloneMovingStats::getStatsForTime(
         FedProDuration timeWindowMillis,
         SteadyTimePoint time)
   {
      size_t bucketCountWindow = timeWindowMillis / _millisPerBucket;
      if (bucketCountWindow < 1) {
         throw std::runtime_error{"Time window is too small."};
      }
      Stats stats{};
      size_t lastCompleteBucket = getBucket(time) - 1;
      {
         std::lock_guard<std::mutex> lock(_guard);
         resetBucketsBeforeUsing(lastCompleteBucket);
         // Copy buckets with current bucket at index n, oldest at index 0
         size_t lastBucketIndex = lastCompleteBucket % _bucketCount;
         size_t oldestBucketIndex = (lastBucketIndex + 1 - bucketCountWindow + _bucketCount) % _bucketCount;
         SampleRangePair maxPerBucket = wrappedRanges(_maxPerBucket, oldestBucketIndex, bucketCountWindow);
         stats._maxSample = max(maxPerBucket);
         SampleRangePair minPerBucket = wrappedRanges(_minPerBucket, oldestBucketIndex, bucketCountWindow);
         stats._minSample = min(minPerBucket);
         SampleRangePair sampleCountPerBucket = wrappedRanges(_sampleCountPerBucket, oldestBucketIndex, bucketCountWindow);
         stats._sampleCount = sum(sampleCountPerBucket);
         stats._bucketCount = bucketCountWindow;
         SampleRangePair sumPerBucket = wrappedRanges(_sumPerBucket, oldestBucketIndex, bucketCountWindow);
         stats._sum = sum(sumPerBucket);
         stats._maxBucket = max(sumPerBucket);
         stats._minBucket = min(sumPerBucket);
         stats._averageBucket = static_cast<float>(stats._sum) / static_cast<float>(bucketCountWindow);
         stats._averageSample =
               stats._sampleCount != 0 ? static_cast<float>(stats._sum) / static_cast<float>(stats._sampleCount) : 0;
         stats._historicTotal = _historicTotal;
         stats._historicCount = _historicCount;
      }
      return stats;
   }

   void StandAloneMovingStats::resetBucketsBeforeUsing(size_t bucketToUse)
   {
      while (bucketToUse >= _nextBucket) {
         _nextBucket = nextBucket(_nextBucket);
         size_t index = _nextBucket % _bucketCount;
         _maxPerBucket[index] = std::numeric_limits<Sample>::min();
         _minPerBucket[index] = std::numeric_limits<Sample>::max();
         _sumPerBucket[index] = 0;
         _sampleCountPerBucket[index] = 0;
      }
   }

   StandAloneMovingStats::SampleRangePair StandAloneMovingStats::wrappedRanges(
         const std::vector<Sample> & src,
         size_t firstElement,
         size_t count) const
   {
      size_t available = _bucketCount - firstElement;
      size_t firstRangeSize = std::min(count, available);
      SampleConstIterator firstRangeBegin{src.begin()};
      std::advance(firstRangeBegin, firstElement);
      SampleConstIterator firstRangeEnd{firstRangeBegin};
      std::advance(firstRangeEnd, firstRangeSize);

      size_t secondRangeSize = count - firstRangeSize;
      SampleConstIterator secondRangeBegin{src.begin()};
      SampleConstIterator secondRangeEnd{secondRangeBegin};
      std::advance(secondRangeEnd, secondRangeSize);

      SampleRange firstRange{firstRangeBegin, firstRangeEnd};
      SampleRange secondRange{secondRangeBegin, secondRangeEnd};
      return {firstRange, secondRange};
   }

   StandAloneMovingStats::Sample StandAloneMovingStats::sum(SampleRangePair ranges)
   {
      Sample m{0};
      for (SampleConstIterator sampleIt{ranges.first.first}; sampleIt != ranges.first.second; ++sampleIt) {
         m += *sampleIt;
      }
      for (SampleConstIterator sampleIt{ranges.second.first}; sampleIt != ranges.second.second; ++sampleIt) {
         m += *sampleIt;
      }
      return m;
   }

   StandAloneMovingStats::Sample StandAloneMovingStats::max(SampleRangePair ranges)
   {
      Sample m{std::numeric_limits<Sample>::min()};
      for (SampleConstIterator sampleIt{ranges.first.first}; sampleIt != ranges.first.second; ++sampleIt) {
         m = std::max(m, *sampleIt);
      }
      for (SampleConstIterator sampleIt{ranges.second.first}; sampleIt != ranges.second.second; ++sampleIt) {
         m = std::max(m, *sampleIt);
      }
      return m;
   }

   StandAloneMovingStats::Sample StandAloneMovingStats::min(SampleRangePair ranges)
   {
      Sample m{std::numeric_limits<Sample>::max()};
      for (SampleConstIterator sampleIt{ranges.first.first}; sampleIt != ranges.first.second; ++sampleIt) {
         m = std::min(m, *sampleIt);
      }
      for (SampleConstIterator sampleIt{ranges.second.first}; sampleIt != ranges.second.second; ++sampleIt) {
         m = std::min(m, *sampleIt);
      }
      return m;
   }

   size_t StandAloneMovingStats::getMaxBucketCount() const
   {
      // The largest possible bucket value that:
      // * Fit within size_t,
      // * allows correct rollover when wrapping, ie (maxBucket + 1) mod _bucketCount = 0
      size_t maxBucket = std::numeric_limits<size_t>::max() - 1;
      // (a + b) % count = (a mod count + b mod count) % count
      maxBucket = maxBucket - ((1 % _bucketCount) + (maxBucket % _bucketCount)) % _bucketCount;
      // Increment, and wraps to zero after maxBucket.
      return maxBucket + 1;
   }

   size_t StandAloneMovingStats::getBucket(SteadyTimePoint time) const
   {
      FedProDuration elapsed = std::chrono::duration_cast<FedProDuration>(time - _startTime);
      return elapsed / _millisPerBucket;
   }

   size_t StandAloneMovingStats::nextBucket(size_t bucket) const
   {
      size_t maxBucketCount = getMaxBucketCount();
      // Increment, and wraps to zero after maxBucket.
      return (1 + (bucket % maxBucketCount)) % maxBucketCount;
   }
}

File: java/session/src/se/pitch/oss/fedpro/common/session/StandAloneMovingStats.java
/*
 *  Copyright (C) 2022 Pitch Technologies AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package se.pitch.oss.fedpro.common.session;

import net.jcip.annotations.GuardedBy;

import java.util.Arrays;

public class StandAloneMovingStats implements MovingStats {
   private static final int DEFAULT_MILLIS_PER_BUCKET = 1_000;

   private final Object _guard = new Object();
   @GuardedBy("_guard")
   private final int[] _maxPerBucket;
   @GuardedBy("_guard")
   private final int[] _minPerBucket;
   @GuardedBy("_guard")
   private final int[] _sumPerBucket;
   @GuardedBy("_guard")
   private final int[] _sampleCountPerBucket;

   private long _historicTotal = 0;
   private long _historicCount = 0;

   private final int _bucketCount;
   private final long _millisPerBucket;
   private long _startTime;

   public StandAloneMovingStats(
         long maxTimeWindowMillis) {
      this(maxTimeWindowMillis, MovingStats.validTimeMillis());
   }

   public StandAloneMovingStats(
         long maxTimeWindowMillis,
         long startTime) {
      this(maxTimeWindowMillis, DEFAULT_MILLIS_PER_BUCKET, startTime);
   }

   public StandAloneMovingStats(
         long maxTimeWindowMillis,
         long millisPerBucket,
         long startTime)
   {
      _millisPerBucket = millisPerBucket;
      _bucketCount = (int)(maxTimeWindowMillis / millisPerBucket) * 2;
      _maxPerBucket = new int[_bucketCount];
      _minPerBucket = new int[_bucketCount];
      _sumPerBucket = new int[_bucketCount];
      _sampleCountPerBucket = new int[_bucketCount];
      reset(startTime);
   }

   @Override
   public void sample(int value) {
      sample(value, MovingStats.validTimeMillis());
   }

   @Override
   public void sample(int value, long now)
   {
      long currentBucket = getBucket(now);
      synchronized (_guard) {
         resetBucketsBeforeUsing(currentBucket);
         int currentBucketIndex = (int) (currentBucket % _bucketCount);
         _maxPerBucket[currentBucketIndex] = Math.max(_maxPerBucket[currentBucketIndex], value);
         _minPerBucket[currentBucketIndex] = Math.min(_minPerBucket[currentBucketIndex], value);
         _sumPerBucket[currentBucketIndex] += value;
         _sampleCountPerBucket[currentBucketIndex]++;
         _historicTotal += value;
         _historicCount++;
      }
   }

   @Override
   public Stats getStats() {
      return getStatsForTime(MovingStats.validTimeMillis());
   }

   @Override
   public Stats getStats(int timeWindowMillis) {
      return getStatsForTime(timeWindowMillis, MovingStats.validTimeMillis());
   }

   public Stats getStatsForTime(long now) {
      return getStatsForTime((int) ((_bucketCount / 2) * _millisPerBucket), now);
   }

   @Override
   public Stats getStatsForTime(int timeWindowMillis, long now)
   {
      int bucketCountWindow = (int)(timeWindowMillis / _millisPerBucket);
      if (bucketCountWindow < 1) {
         throw new RuntimeException("Time window is too small.");
      }
      Stats stats = new Stats();
      int[] maxPerBucket;
      int[] minPerBucket;
      int[] sampleCountPerBucket;
      int[] sumPerBucket;
      long lastCompleteBucket = getBucket(now) - 1;
      synchronized (_guard) {
         resetBucketsBeforeUsing(lastCompleteBucket);
         // Copy buckets with current bucket at index n, oldest at index 0
         int lastBucketIndex = (int) (lastCompleteBucket % _bucketCount);
         int oldestBucketIndex = (lastBucketIndex + 1 - bucketCountWindow + _bucketCount) % _bucketCount;
         maxPerBucket = copyWrappedArray(_maxPerBucket, oldestBucketIndex, bucketCountWindow);
         minPerBucket = copyWrappedArray(_minPerBucket, oldestBucketIndex, bucketCountWindow);
         sampleCountPerBucket = copyWrappedArray(_sampleCountPerBucket, oldestBucketIndex, bucketCountWindow);
         sumPerBucket = copyWrappedArray(_sumPerBucket, oldestBucketIndex, bucketCountWindow);
         stats.historicTotal = _historicTotal;
         stats.historicCount = _historicCount;
      }
      int sampleCount = sum(sampleCountPerBucket);
      stats.sampleCount = sampleCount;
      stats.bucketCount = bucketCountWindow;

      stats.maxSample = max(maxPerBucket);
      stats.minSample = min(minPerBucket);
      stats.sum = sum(sumPerBucket);
      stats.maxBucket = max(sumPerBucket);
      stats.minBucket = min(sumPerBucket);
      stats.averageBucket = (float) stats.sum / bucketCountWindow;
      stats.averageSample = sampleCount != 0 ? (float) stats.sum / sampleCount : 0;
      return stats;
   }

   @GuardedBy("_guard")
   private long _nextBucket = 0;

   @GuardedBy("_guard")
   private void resetBucketsBeforeUsing(long bucketToUse) {
      while (bucketToUse >= _nextBucket) {
         _nextBucket = nextBucket(_nextBucket);
         int index = (int) (_nextBucket % _bucketCount);
         _maxPerBucket[index] = Integer.MIN_VALUE;
         _minPerBucket[index] = Integer.MAX_VALUE;
         _sumPerBucket[index] = 0;
         _sampleCountPerBucket[index] = 0;
      }
   }

   private int[] copyWrappedArray(int[] src, int firstElement, int count)
   {
      int[] dest = new int[count];
      int available = _bucketCount - firstElement;
      int firstRange = Math.min(count, available);
      System.arraycopy(src, firstElement, dest, 0, firstRange);

      int secondRange = count - firstRange;
      if (count != 0) {
         System.arraycopy(src, 0, dest, firstRange, secondRange);
      }
      return dest;
   }

   private static int sum(int... buckets)
   {
      int m = 0;
      for (int n : buckets) {
         m += n;
      }
      return m;
   }

   private static int max(int... buckets)
   {
      int m = Integer.MIN_VALUE;
      for (int n : buckets) {
         m = Math.max(m, n);
      }
      return m;
   }

   private static int min(int... buckets)
   {
      int m = Integer.MAX_VALUE;
      for (int n : buckets) {
         m = Math.min(m, n);
      }
      return m;
   }


   @GuardedBy("_guard")
   private void reset(long startTime)
   {
      Arrays.fill(_maxPerBucket, Integer.MIN_VALUE);
      Arrays.fill(_minPerBucket, Integer.MAX_VALUE);
      Arrays.fill(_sumPerBucket, 0);
      Arrays.fill(_sampleCountPerBucket, 0);
      _startTime = startTime;
   }

   public long getBucketCount()
   {
      return _bucketCount;
   }

   long getMaxBucketCount()
   {
      // Compute the largest possible bucket value that:
      // * Fit within int,
      // * allows correct rollover when wrapping, ie (maxBucket + 1) mod _bucketCount = 0
      long maxBucket = Long.MAX_VALUE - 1;
      // (a + b) % count = (a mod count + b mod count) % count
      maxBucket = maxBucket - (1 + (maxBucket % _bucketCount)) % _bucketCount;
      return maxBucket + 1;
   }

   long getBucket(long now)
   {
      long elapsed = now - _startTime;
      return (elapsed / _millisPerBucket) % getMaxBucketCount();
   }

   long nextBucket(long bucket)
   {
      long maxBucketCount = getMaxBucketCount();
      // Increment, and wraps to zero after maxBucket.
      return (1 + (bucket % maxBucketCount)) % maxBucketCount;
   }

}
